Update `jnp.floor_divide` to make it consistent with `np.floor_divide` for division by zero

rajasekharporeddy opened the pull request:
Fixes #11990

Current behaviour:
```python
>>> jnp.floor_divide(2, 0)
Array(-2, dtype=int32, weak_type=True)
>>> np.floor_divide(2, 0)
<stdin>:1: RuntimeWarning: divide by zero encountered in floor_divide
0
```
With thi PR:
```python
>>> jnp.floor_divide(2, 0)
Array(0, dtype=int32, weak_type=True)
```

jakevdp commented on jax/_src/numpy/ufuncs.py:
```diff
@@ -2495,12 +2495,20 @@ def floor_divide(x1: ArrayLike, x2: ArrayLike, /) -> Array:
   dtype = dtypes.dtype(x1)
   if dtypes.issubdtype(dtype, np.unsignedinteger):
-    return lax.div(x1, x2)
+    result = lax.div(x1, x2)
+    result = lax.select(_broadcast_to(x2 == 0, result.shape),
+                        lax.full_like(result, fill_value=0),
+                        result)
+    return result
```
Using `_where` would make this more succinct. I think this should work:
```suggestion
    return _where(x2 == 0, 0, lax.div(x1, x2))
```

rajasekharporeddy replied:
Thank! Done.

jakevdp commented on jax/_src/numpy/ufuncs.py:
```diff
@@ -2495,12 +2495,20 @@ def floor_divide(x1: ArrayLike, x2: ArrayLike, /) -> Array:
     select = logical_and(lax.sign(x1) != lax.sign(x2), lax.rem(x1, x2) != 0)
     # TODO(mattjj): investigate why subtracting a scalar was causing promotion
-    return _where(select, quotient - 1, quotient)
+    result = _where(select, quotient - 1, quotient)
+    result = lax.select(_broadcast_to(x2 == 0, result.shape),
+                        lax.full_like(result, fill_value=0),
+                        result)
+    return result
```
Similarly here:
```suggestion
    return _where(x2 == 0, 0, _where(select, quotient - 1, quotient)
```

jakevdp commented on jax/_src/numpy/ufuncs.py:
```diff
@@ -2559,6 +2567,10 @@ def _float_divmod(x1: ArrayLike, x2: ArrayLike) -> tuple[Array, Array]:
   mod = lax.select(ind, mod + x2, mod)
   div = lax.select(ind, div - _constant_like(div, 1), div)
 
+  div = lax.select(_broadcast_to(x2 == 0, div.shape),
+                   lax.full_like(div, fill_value=np.inf) * lax.sign(x1),
+                   div)
+
```
We need to double check this logic for when `x2` is `-0.0` (the sign of the returned infinity may be wrong), or when `x1` and `x2` are both zero (when the result should be NaN).

rajasekharporeddy replied:
Thanks! Modified the `_float_divmod` logic to consider all the possible scenarios to match with NumPy.

jakevdp commented on jax/_src/numpy/ufuncs.py:
```diff
@@ -2557,8 +2557,14 @@ def _float_divmod(x1: ArrayLike, x2: ArrayLike) -> tuple[Array, Array]:
   div = lax.select(ind, div - _constant_like(div, 1), div)
 
+  div = _where(x2 == 0., lax.div(x1, x2), div)
+  negative_inputs = logical_and(lax.sign(x1)==-1, lax.sign(x2)==-1)
```
We should use `signbit` rather than `sign`, because this will properly handle signed zeros and NaNs.

rajasekharporeddy replied:
Thanks! Done.

jakevdp commented on jax/_src/numpy/ufuncs.py:
```diff
@@ -2557,8 +2557,14 @@ def _float_divmod(x1: ArrayLike, x2: ArrayLike) -> tuple[Array, Array]:
+  div = _where(x2 == 0., lax.div(x1, x2), div)
+  negative_inputs = logical_and(lax.sign(x1)==-1, lax.sign(x2)==-1)
+  div = _where(logical_and(lax.gt(x1, x2), negative_inputs), negative(div), div)
+  div = _where(x1 == 0., div * lax.sign(x1), div)
```
If we use `signbit` above then I think this may not be necessary

rajasekharporeddy replied:
Thanks! Done.

jakevdp reviewed (commented):
A couple more comments below.

Also, it would be worth adding some additional testing of specific cases that this is meant to handle.

rajasekharporeddy commented:
Added the testing for divide by zero for both integer and float.

jakevdp commented on tests/lax_numpy_operators_test.py:
```diff
@@ -731,6 +731,29 @@ def testSpacingSubnormals(self, dtype):
+    x1 = jnp.array([-2., 3., -4., 6., -0., 2., -4., 0., -2., 0., -0., 0.], dtype=float)
+    x2 = jnp.array([-3., -2., 3., -7., 1., -0., 0., -2., 1., 0., 0., -0.], dtype=float)
+    jnp_result = jnp.floor_divide(x1, x2)
+    np_floor_divide = jtu.ignore_warning(category=RuntimeWarning,
```
It's a bit clearer I think to put these `ignore_warning` statements as decorators for the whole test function.

rajasekharporeddy replied:
Thanks for the suggestion! Included `ignore_warning` statements as decorators.

rajasekharporeddy commented on tests/lax_numpy_operators_test.py:
```diff
@@ -731,6 +731,29 @@ def testSpacingSubnormals(self, dtype):
+
+  def testFloorDivideByZeroFloat(self):
+    x1 = jnp.array([-2., 3., -4., 6., -0., 2., -4., 0., -2., 0., -0., 0.], dtype=float)
+    x2 = jnp.array([-3., -2., 3., -7., 1., -0., 0., -2., 1., 0., 0., -0.], dtype=float)
```
Thanks for the suggestion! Done

jakevdp commented on tests/lax_numpy_operators_test.py:
```diff
@@ -731,6 +731,30 @@ def testSpacingSubnormals(self, dtype):
     self._CompileAndCheck(jnp.spacing, args_maker, tol=0)
 
+  @jtu.ignore_warning(category=RuntimeWarning,
+                      message="divide by zero.*")
```
nit: `message` could go on the same line to make this easier to read

rajasekharporeddy replied:
Thanks! Done.

rajasekharporeddy commented:
Squashed the commits into one.

jakevdp commented on jax/_src/numpy/ufuncs.py:
```diff
@@ -2555,10 +2555,15 @@ def _float_divmod(x1: ArrayLike, x2: ArrayLike) -> tuple[Array, Array]:
   div = lax.div(lax.sub(x1, mod), x2)
 
-  ind = lax.bitwise_and(mod != 0, lax.sign(x2) != lax.sign(mod))
+  ind = lax.bitwise_and(mod != 0, signbit(x2) != signbit(mod))
```
I think we could stick with `lax.sign` here, since we're looking at non-zero values (and it lowers more directly to efficient XLA ops).

rajasekharporeddy replied:
Done.

jakevdp commented on jax/_src/numpy/ufuncs.py:
```diff
@@ -2555,10 +2555,15 @@ def _float_divmod(x1: ArrayLike, x2: ArrayLike) -> tuple[Array, Array]:
-  ind = lax.bitwise_and(mod != 0, lax.sign(x2) != lax.sign(mod))
+  ind = lax.bitwise_and(mod != 0, signbit(x2) != signbit(mod))
   mod = lax.select(ind, mod + x2, mod)
+  mod = _where(signbit(x2) != signbit(mod), negative(mod), mod)
```
Can we use `mod = copysign(mod, x2)` here?

Also, here and below, it's slightly more efficient to use `lax.neg` rather than `negative`; the latter does a bunch of validation that has already been done in this function.

rajasekharporeddy replied:
Thanks! `copysign` works here. replaced `negative` with `lax.neg` in the other place.

jakevdp reviewed (changes requested):
Looks great! Now that we have comprehensive tests, a few small comments on the implementation:

jakevdp reviewed (commented):
Looks great! Please squash the commits, and then we can get this merged 😁

rajasekharporeddy commented:
Squashed

jakevdp commented on tests/lax_numpy_operators_test.py:
```diff
@@ -731,6 +731,26 @@ def testSpacingSubnormals(self, dtype):
+    jnp_result = jnp.floor_divide(x1, x2)
+    np_result = np.floor_divide(x1, x2)
+    self.assertAllClose(jnp_result, np_result, check_dtypes=True)
+    self.assertAllClose(jnp.signbit(jnp_result), jnp.signbit(np_result), check_dtypes=True)
```
It appears that this assertion is failing on some platforms. I'm not sure why:
```
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 2 / 100 (2%)
 ACTUAL: array([[ True,  True,  True,  True,  True,  True,  True,  True,  True,
         True],
       [False, False, False, False, False,  True, False, False, False,...
 DESIRED: array([[ True,  True,  True,  True,  True,  True,  True,  True,  True,
         True],
       [False, False, False, False, False,  True, False, False, False,...
```

rajasekharporeddy replied:
I think the problem, here is the usage of `jnp.signbit` to test the sign of the result. In NumPy, the signbit of `nan` resulted in some cases is set even though the signbit of both the inputs is not set.

```python
>>> np.signbit([np.inf, 1.])
array([False, False])
>>> np.signbit(np.floor_divide(np.inf, 1.))
<ipython-input-52-d8652a8e4b0b>:1: RuntimeWarning: invalid value encountered in floor_divide
  np.signbit(np.floor_divide(np.inf, 1.))
True
>>> jnp.signbit(floor_divide(np.inf, 1.))
Array(False, dtype=bool)
```
We can avoid this by using `lax.sign` instead of `jnp.signbit`.

jakevdp replied:
Changing to `lax.sign` will not correctly differentiate between `-0.0` and `0.0`, or between `-NaN` and `NaN`, which as I understand it is the only point of this check. If we change to `lax.sign`, then this test becomes unnecessary, because the sign match is already covered by the equality check one line above.

So I'd say, either we figure out how to make the `signbit` version work, or we remove the check altogether. What do you think?

rajasekharporeddy replied:
I could not find a way to resolve this. I have tested it on colab CPU and Mac M1 Pro chip. The tests does not fail in these environments with NumPy 1.26.4, 2.1.3 and 2.2.0rc1 versions.

rajasekharporeddy commented:
Hi @jakevdp, Removed the test on `signedness of 0 and NaN` and simplified the `floor_divide` logic. The tests are now written only to test for the `divide by 0` case.

jakevdp commented on tests/lax_numpy_operators_test.py:
```diff
@@ -731,6 +731,24 @@ def testSpacingSubnormals(self, dtype):
+  @jtu.ignore_warning(category=RuntimeWarning, message="invalid value.*")
+  def testFloorDivideFloatSpecialCases(self):
+    x1 = jnp.array([-np.nan, -np.inf, -2., -1., -0., 0., 1., 2., np.inf, np.nan])
+    x2 = 0.
```
Why not test all the special values in `x2` as well?

rajasekharporeddy replied:
Thanks! Test case is modified so as to test for all the specific values in `x2`.

jakevdp commented on tests/lax_numpy_operators_test.py:
```diff
@@ -731,6 +731,26 @@ def testSpacingSubnormals(self, dtype):
+
+  @jtu.sample_product(
+      x1 = [-np.nan, -np.inf, -2., -1., -0., 0., 1., 2., np.inf, np.nan],
+      x2 = [-np.nan, -np.inf, -2., -1., -0., 0., 1., 2., np.inf, np.nan],
```
The largest number of cases we regularly test is 90, this sampling is 100 cases, meaning that not all will be tested in practice.

Instead I'd recommend a single test where you define `x = jnp.array([-np.nan, -np.inf...])` and then check the output of `jnp.floor_divide(x[:, None], x[None, :])`

rajasekharporeddy replied:
Thank! Done.

jakevdp commented:
I tracked it down to this line: https://github.com/jax-ml/jax/blob/2b06f93c703d062bebb6154a0dc030f2467c67cc/jax/_src/lax/control_flow/loops.py#L383

This causes all characters after the first within the error message to be converted to lowercase, which results in the error output you're seeing.

rajasekharporeddy commented:
Hi @jakevdp, I have modified the code to avoid the tests failures in TPU. Could you please verify and approve it again.

rajasekharporeddy committed:
Update jnp.floor_divide to make it consistent with np.floor_divide for division by zero

jakevdp commented on jax/_src/numpy/ufuncs.py:
```diff
@@ -2553,11 +2553,14 @@ def divmod(x1: ArrayLike, x2: ArrayLike, /) -> tuple[Array, Array]:
   ind = lax.bitwise_and(mod != 0, lax.sign(x2) != lax.sign(mod))
   mod = lax.select(ind, mod + x2, mod)
   div = lax.select(ind, div - _constant_like(div, 1), div)
+  if dtypes.dtype(x1) != dtypes.bfloat16:
```
This is surprising – can you comment on why we need specialized logic for bfloat16 here?

rajasekharporeddy replied:
Tests are getting failed when `x1` is of `bfloat16` type and `x2` is of `int` or `uint` type. Hence I imposed a condition on it. But even then some of the tests are failing.

jakevdp replied:
I would remove this if-else block and figure out how to fix the test failure without having the implementation branch on the dtype. If this branching is actually necessary, we should add a comment about why.

jakevdp commented on jax/_src/numpy/ufuncs.py:
```diff
@@ -2553,11 +2553,14 @@ def divmod(x1: ArrayLike, x2: ArrayLike, /) -> tuple[Array, Array]:
 def _float_divmod(x1: ArrayLike, x2: ArrayLike) -> tuple[Array, Array]:
   # see float_divmod in floatobject.c of CPython
   mod = lax.rem(x1, x2)
+  mod = _where(logical_and(abs(x2) == np.inf, abs(x1) != np.inf), x1, mod)
```
`isinf` would be better here:
```suggestion
  mod = _where(isinf(x2) & ~isinf(x1), x1, mod)
```

rajasekharporeddy replied:
Done.
